### PR TITLE
Initial implementation of DRCOV-compatible code coverage collection.

### PR DESCRIPTION
**Pull request by assafcarlsbad**
- The output file can be loaded by an IDA plugin such as Lighthouse in order to visualize the collected code coverage.
image
- For this revision, only the UEFI loader was modified to keep a list of loaded images. I believe this amendment can be ported to the other loaders with relative ease.

### CODE REVIEW

**Comment by xwings** [comment]
Hi,

Add your name into credits.txt too.

**Comment by assafcarlsbad** [comment]
Already added 😄

**Comment by xwings** [comment]
yo,

I just woke up and on my mobile. So i did not realize it,

Btw, qiling/coverage.py  need to be somewhere. It should not be in qiling. This is a feature not widely use in the core engine.

**Comment by aquynh** [comment]
Please put more docs in coverage.py and qltool that this is DRCOV format, from DynamoRio

Also, making it easy to extend for other coverage formats in the future will be nice.

**Comment by assafcarlsbad** [comment]
Hi, I made some changes in accordance with your comments. Can you please re-review?

**Comment by aquynh** [comment]
Nice work! 

Does qltool report error if i pass in an invalid coverage name?

**Comment by aquynh** [comment]
Please update README with a new example.

**Comment by assafcarlsbad** [comment]
@aquynh 

> Nice work!
> 
> Does qltool report error if i pass in an invalid coverage name?

```
carlsbad@AssafC-T490:/mnt/c/Users/carlsbad/Code/qiling$ ./qltool run --rootfs . --debug --coverage-format foo -c new_tcg_dxe.cov -h /mnt/c/Users/carlsbad/Code/UEFI_RETool/modules/TcgDxe /mnt/c/Users/carlsbad/Code/UEFI_RETool/modules/TcgPlatformSetupPolicy
usage: qltool run [-h] [-f FILE] --rootfs ROOTFS [--args ...] [-o OUTPUT] [-v VERBOSE] [--env FILE] [-g GDB] [--profile PROFILE]
                  [--strace] [--dump] [--debug] [--disasm] [--console CONSOLE] [-e FUNCTION NAME] [--log-dir DIRECTORY NAME]
                  [--trace] [--root] [--debug_stop] [-m] [--timeout TIMEOUT] [-c COVERAGE_FILE] [--coverage-format {drcov}]
                  ...
qltool run: error: argument --coverage-format: invalid choice: 'foo' (choose from 'drcov')
```

> Please update README with a new example.

Done :)

File: README.md [hunk 1]
```diff
@@ -166,6 +166,11 @@ $ ./qltool run -f examples/rootfs/x8664_linux/bin/x8664_hello --gdb 127.0.0.1:99
 
 See  https://docs.qiling.io/  for more details
 
+With code coverage collection (UEFI only for now):
+
+```
+$ ./qltool run -f TcgDxe.efi --rootfs . --coverage-format drcov --coverage-file TcgDxe.cov
```
Comment (xwings) [inline README.md:172]: is should be ,./qltool run -f examples/rootfs/xxxx

Comment (assafcarlsbad) [reply README.md]: Fixed.

**Comment by jershmagersh** [comment]
Great work, I was thinking of implementing the same thing, but you've done a much better job. This should also work with something like DragonDance: https://github.com/0ffffffffh/dragondance

**Comment by aquynh** [comment]
A final request from me: please have a README in coverage/ about this coverage framework, with some intructions on how to extend it for a new coverage.

**Comment by xwings** [comment]
> A final request from me: please have a README in coverage/ about this coverage framework, with some intructions on how to extend it for a new coverage

the readme should goes to docs, maybe a differents pr in https://github.com/qilingframework/rtfd.io

**Comment by aquynh** [comment]
it is good to have that in source code, too.

**Comment by assafcarlsbad** [comment]
Done. Please let me know if I'm missing something or if you want me to elaborate more on some aspects.

**Comment by xwings** [comment]
perfect ! 

thanks !

**Pull request merged**